src/drivers/virtio/virtqueue/mod.rs

```diff
@@ -512,27 +512,65 @@ pub enum BufferType {
 	Indirect,
 }
 
-/// MemPool allows to easily control, request and provide memory for Virtqueues.
-///
-/// The struct is initialized with a limit of free running "tracked"
-/// memory descriptor ids. As Virtqueus do only allow a limited amount of descriptors in their queue,
-/// the independent queues, can control the number of descriptors by this.
-struct MemPool {
-	pool: Vec<u16>,
-	limit: u16,
-}
+mod index_alloc {
+	use alloc::boxed::Box;
+
+	use align_address::Align;
 
-impl MemPool {
-	/// Returns a given id to the id pool
-	fn ret_id(&mut self, id: u16) {
-		self.pool.push(id);
+	/// This type allows allocating indices.
+	///
+	/// The indices can be used as descriptor IDs.
+	pub struct IndexAlloc {
+		/// Zero bits are available.
+		bits: Box<[usize]>,
 	}
 
-	/// Returns a new instance, with a pool of the specified size.
-	fn new(size: u16) -> MemPool {
-		MemPool {
-			pool: (0..size).collect(),
-			limit: size,
+	const USIZE_BITS: usize = usize::BITS as usize;
+
+	impl IndexAlloc {
+		pub fn new(len: usize) -> Self {
+			let usizes = len.align_up(USIZE_BITS) / USIZE_BITS;
+			let extra_bits = len % USIZE_BITS;
+
+			let mut bits = vec![0; usizes].into_boxed_slice();
+
+			if extra_bits != 0 {
+				*bits.last_mut().unwrap() |= usize::MAX >> extra_bits;
+			}
+
+			Self { bits }
+		}
+
+		#[inline]
+		pub fn allocate(&mut self) -> Option<usize> {
+			for (word_index, word) in self.bits.iter_mut().enumerate() {
+				let trailing_ones = word.trailing_ones();
+				if trailing_ones < usize::BITS {
+					let mask = 1 << trailing_ones;
+					*word |= mask;
+					let index = word_index * USIZE_BITS + usize::try_from(trailing_ones).unwrap();
+					return Some(index);
+				}
+			}
+
+			None
```

Comment:
```suggestion
			let (word_index, trailing_ones) = self
				.bits
				.iter()
				.copied()
				.map(usize::trailing_ones)
				.enumerate()
				.find(|(_, trailing_ones)| *trailing_ones < usize::BITS)?;

			let mask = 1 << trailing_ones;
			self.bits[word_index] |= mask;
			let index = word_index * USIZE_BITS + usize::try_from(trailing_ones).unwrap();
			Some(index)
```
I am not sure if it would be an improvement but wanted to offer it as an option. It would save us from some nesting.